torch/_inductor/lowering.py

```diff
@@ -1853,18 +1854,43 @@ def fn(idx):
     )
 
 
+# Note [Decompositions as Inductor Lowerings]
 # This is moved from decomposition to lowering because this decomp introduced
 # mutation in the graph, which is bad for Aot Autograd. Aot Autograd runs dead
 # code elimination and common subexpression elimination optimizations, which
 # assume graphs to be side-effect free. More details at
 # https://github.com/pytorch/torchdynamo/issues/1235.
 # Moving such reinplacing type of decomps to lowering ensures that AotAutograd
 # gets good graphs.
+# A better long term solution would be to functionalize all decompositions.
+# That would let us use these decompositions "normally",
+# instead of forcing them to be lowerings.
 @register_lowering([aten.index_put])
 def index_put(x, indices, values, accumulate=False):
     return index_put_(clone(x), indices, values, accumulate)
 
 
+# See Note [Decompositions as Inductor Lowerings]
+# We don't want the decomp to do type promotion implicitly;
+# It will try to convert index (int64) to a float-like type.
+@register_lowering([aten.index_add], type_promotion_kind=None)
+def index_add(x, dim, index, tensor, *, alpha=1):
+    return index_add_(clone(x), dim, index, tensor, alpha=alpha)
+
+
+# See Note [Decompositions as Inductor Lowerings]
+# We don't want the decomp to do type promotion implicitly;
+# It will try to convert index (int64) to a float-like type.
+@register_lowering([aten.index_add_], type_promotion_kind=None)
+def index_add_(x, dim, index, tensor, *, alpha=1):
+    dim = _prims_common.canonicalize_dims(len(x.get_size()), dim)
+    if alpha != 1:
+        tensor = mul(tensor, alpha)
+    idx = (None,) * dim + (index,)
+    index_put_(x, idx, tensor, accumulate=True)
+    return x
```

Comment:
@ngimel are there any known silent correctness issues we should test this against? It looks like `index_add()` decomposes into `index_add_()`, which was getting silently DCE'd.

Also, looking for feedback from someone familiar with inductor! I basically copied the decomps from `torch/_decomp/decompositions.py`, and tried to replace aten op calls with lowering calls.

I also had to set `type_promotion_kind=None`, since it looks like otherwise inductor will implicitly type promote the inputs (and convert the integral index tensor into a float)

Reply:
Yeah `index_add_` issue was discovered in https://github.com/pytorch/torchdynamo/issues/1863, does this PR contain a fix?

Reply:
It should. I basically adopted the same strategy inductor used for `index_put()`, and moved the decomp to a lowering. That way `index_put()` isn't decomposed in aot autograd, and instead it gets decomposed internally in inductor.


Tbh I think the better long term fix would be to functionalize all of our decomps (specifically, if a decomp running inside of proxy_tensor decomposes into mutable ops, we should "re-functionalize it" directly inside of proxy tensor). But this is a quicker workaround just so that I can successfully add the assert in aot autograd mentioned in this PR.

Reply:
There's another fix needed here, from #87183 https://github.com/pytorch/pytorch/pull/87183/files#diff-d25ad317caf41c3fdf1de3bcfc730102c9e4a3e68639559017ae45cb0bc26911R1747. 
Also, what about alpha dtype and index ndim checks currently in decomp? Would they happen somewhere else instead? 
I agree with the better long term fix, but it's fine to sidestep it for now. Setting `type_promotion_kind` to None is right.

Reply:
Oh wait, maybe that fix isn't needed if inductors `index_put_` is more permissive and would accept 0d `x`, I don't know off the top of my head